### PR TITLE
PCC05 Geoff Riley

### PR DESCRIPTION
**Pull request by GeoffRiley**
Difficulty level (1-10): [7]
Estimated time spent (hours): [8]
Completed (yes/no): [Yes]
I stretched my coding skills (if yes what did you learn?): [nltk and gensim libraries… frightfully complicated and seemingly impossible to know if you've got a 'right' solution!]
Other feedback (what can we improve?): []

### CODE REVIEW

File: 05/GeoffRiley/similar_tweeters.py [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+# exclusions:
+#  1. Tags starting with '#'
+#  2. email addresses containing '@'
+#  3. web addresses containing 'http' or 'https'
+#  4. URL indicators beginning '//'
+#  5. numbers comprising one or more digits followed by an optional two alphas
+exclusions = re.compile(r'^(#.*|.*@.*|.*https?|//.*|\d+([a-zA-Z]{2})?)$')
+my_punct = ''.join(x for x in string.punctuation if x not in ['#', '@'])
+trans = str.maketrans(my_punct, ' ' * len(my_punct))
```
Comment (bbelderbos) [inline 05/GeoffRiley/similar_tweeters.py]: `maketrans` is pretty cool

File: 05/GeoffRiley/similar_tweeters.py [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def clean_words(usr: UserTweets) -> list:
+    result = []
+    for twitter in usr:
+        for word in twitter.text.translate(trans).lower().split():
+            if word not in sw and len(word) > 3 and not exclusions.match(word):
```
Comment (bbelderbos) [inline 05/GeoffRiley/similar_tweeters.py]: got a bit confused where `sw` was defined, I would name it `STOPWORDS` = more descriptive and uppercase makes me go to the top of the module

Comment (bbelderbos) [reply 05/GeoffRiley/similar_tweeters.py]: you could make `3` a constant as well: `MIN_CHARS_WORD` or is that overkill?

Comment (bbelderbos) [reply 05/GeoffRiley/similar_tweeters.py]: would this exclude the #AI hashtag btw?

File: 05/GeoffRiley/similar_tweeters.py [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+def similar_tweeters(user1, user2):
+    tw1 = clean_words(UserTweets(user1))
+    tw2 = clean_words(UserTweets(user2))
+    dictionary = corpora.Dictionary([tw1, tw2])
+    vector = [dictionary.doc2bow(tw1), dictionary.doc2bow(tw2)]
+    lsi = models.LsiModel(vector, id2word=dictionary, num_topics=100)
+    vec_lsi = lsi[vector]
+
+    index = similarities.MatrixSimilarity(vec_lsi)
```
Comment (bbelderbos) [inline 05/GeoffRiley/similar_tweeters.py]: cool stuff

**Comment by bbelderbos** [comment]
good stuff!

File: 05/GeoffRiley/similar_tweeters.py (hunk 1 above)
Comment (GeoffRiley) [reply 05/GeoffRiley/similar_tweeters.py]: It's something I knew about but haven't used before… I know that there is a lot more flexibility to it and will be having a play.

File: 05/GeoffRiley/similar_tweeters.py (hunk 2 above)
Comment (GeoffRiley) [reply 05/GeoffRiley/similar_tweeters.py]: Yes, it is killing all hash tags… I don't really know how *significant* they are as part of a dialog.

File: 05/GeoffRiley/similar_tweeters.py (hunk 3 above)
Comment (GeoffRiley) [reply 05/GeoffRiley/similar_tweeters.py]: It's very difficult to work out just how this library is supposed to be used without going through all the maths behind it and I think that'll take me quite a long time to digest! :grinning:

**Pull request merged**